File: RemoveNthNodeFromEndofList.cpp
/*
Given the head of a linked list, remove the nth node from the end of the list and return its head.
  
Example:
Input: head = [1,2,3,4,5], n = 2
Output: [1,2,3,5]
Example 2:

Input: head = [1], n = 1
Output: []
Example 3:

Input: head = [1,2], n = 1
Output: [1]
  */

 /**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* end = head;
        ListNode* nMinusOneThFromEnd = head;

        if(!head)
            return nullptr;

        while(end && n){
            end = end->next;
            n--;
        }

        if(n)
            return nullptr;

        if(end){  
            end = end->next;  
        }
        else{
            end = head;
            cout<<end->val;
            head = head->next;
            delete end;
            return head;
        }

        while(end)
        {
            nMinusOneThFromEnd = nMinusOneThFromEnd->next;
            end = end->next;
        }
        
        end = nMinusOneThFromEnd->next;
        nMinusOneThFromEnd->next = nMinusOneThFromEnd->next->next;
        delete end;

        return head;
    }
};

File: camp_I/week1/leetcode/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        endNode = dummy = ListNode(0, head)     
        i = 1
        while i <= n:
            endNode = endNode.next
            i += 1
        
        beforNthNode = dummy
        while endNode.next:
            endNode = endNode.next
            beforNthNode = beforNthNode.next
        
        beforNthNode.next = beforNthNode.next.next

        return dummy.next
        
